File: app/ios/Runner/TextRecogniserPlugin.swift
import Flutter
import UIKit
import Vision

/// A one-shot latch, so a Flutter reply can never be sent twice.
private final class Replied {
    private var done = false
    private let lock = NSLock()

    /// Returns true exactly once.
    func claim() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if done { return false }
        done = true
        return true
    }
}

/// On-device text recognition using Apple's Vision framework.
///
/// Vision rather than ML Kit deliberately: ML Kit's iOS pods ship no arm64
/// simulator slice, so depending on them costs the whole team iOS simulator
/// development. Vision is first-party, free, on-device, faster, and has no
/// such limitation. See ADR-0004.
///
/// This class recognises text and nothing else. Choosing which digit run is
/// the meter register happens in Dart, in `DigitExtractor`, so that judgement
/// stays testable without a device and identical on both platforms.
final class TextRecogniserPlugin: NSObject, FlutterPlugin {

    private static let channelName = "grid/text_recogniser"

    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(TextRecogniserPlugin(), channel: channel)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "isAvailable":
            // VNRecognizeTextRequest has been available since iOS 13, and the
            // deployment target is above that — but report it honestly rather
            // than asserting it.
            if #available(iOS 13.0, *) {
                result(true)
            } else {
                result(false)
            }

        case "recognise":
            guard
                let args = call.arguments as? [String: Any],
                let path = args["path"] as? String
            else {
                result(FlutterError(
                    code: "bad_arguments",
                    message: "recognise requires a 'path'",
                    details: nil
                ))
                return
            }
            recognise(path: path, rawResult: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func recognise(path: String, rawResult: @escaping FlutterResult) {
        // FlutterResult must be called exactly once, on the platform thread.
        //
        // Vision runs its completion handler on whatever queue performed the
        // request — a background one here — and calling back into Flutter from
        // there is undefined behaviour. Worse, `perform` can throw *after* the
        // completion handler has already fired, which would reply twice and
        // trip Flutter's "reply already submitted" fatal error.
        //
        // This wrapper hops to main and drops everything after the first call.
        let replied = Replied()
        let result: FlutterResult = { value in
            guard replied.claim() else { return }
            if Thread.isMainThread {
                rawResult(value)
            } else {
                DispatchQueue.main.async { rawResult(value) }
            }
        }

        guard
            let image = UIImage(contentsOfFile: path),
            let cgImage = image.cgImage
        else {
            result(FlutterError(
                code: "unreadable_image",
                message: "Could not load an image at \(path)",
                details: nil
            ))
            return
        }

        guard #available(iOS 13.0, *) else {
            result([])
            return
        }

        let request = VNRecognizeTextRequest { request, error in
            if error != nil {
                result([])
                return
            }
            guard let observations = request.results as? [VNRecognizedTextObservation] else {
                result([])
                return
            }
            result(observations.compactMap(Self.encode))
        }

        // A meter register is a short run of digits in a known style. Accurate
        // beats fast here: the budget is 1500 ms and recognition of a single
        // cropped frame lands well inside it, while `.fast` measurably confuses
        // 8 with 0 on a mechanical register.
        request.recognitionLevel = .accurate

        // Language correction actively hurts. It is tuned for words, and it
        // will happily turn "045821" into something it likes better.
        request.usesLanguageCorrection = false

        // A meter face carries no prose. Restricting the character set removes
        // whole classes of misread — an 'S' where a 5 is, a 'B' where an 8 is.
        if #available(iOS 16.0, *) {
            request.revision = VNRecognizeTextRequestRevision3
        }

        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                try handler.perform([request])
            } catch {
                result([])
            }
        }
    }

    /// Vision reports a bounding box in a bottom-left origin, normalised to
    /// the image. Dart expects a top-left origin, so the y axis flips here —
    /// getting this wrong makes every candidate score as if it sat at the
    /// opposite edge of the face, and the extractor quietly picks the serial
    /// number instead.
    @available(iOS 13.0, *)
    private static func encode(_ observation: VNRecognizedTextObservation) -> [String: Any]? {
        guard let candidate = observation.topCandidates(1).first else { return nil }
        let box = observation.boundingBox
        return [
            "text": candidate.string,
            "confidence": Double(candidate.confidence),
            "left": Double(box.origin.x),
            "top": Double(1.0 - box.origin.y - box.size.height),
            "width": Double(box.size.width),
            "height": Double(box.size.height),
        ]
    }
}

File: app/android/app/src/main/kotlin/com/gridapp/grid/TextRecogniserPlugin.kt
package com.gridapp.grid

import android.graphics.BitmapFactory
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/**
 * On-device text recognition using ML Kit.
 *
 * ML Kit is the right engine on Android — it is bundled, accelerated and
 * accurate. It is only unusable on *iOS*, where its pods ship no arm64
 * simulator slice, which is why the two platforms use different engines
 * behind one Dart interface rather than sharing a Flutter package. See
 * ADR-0004.
 *
 * This class recognises text and nothing else. Choosing which digit run is
 * the meter register happens in Dart, in `DigitExtractor`, so that judgement
 * stays testable without a device and identical on both platforms.
 */
class TextRecogniserPlugin(
    private val context: android.content.Context,
) : MethodChannel.MethodCallHandler {

    companion object {
        const val CHANNEL = "grid/text_recogniser"
    }

    private val recogniser by lazy {
        TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "isAvailable" -> result.success(true)
            "recognise" -> recognise(call, result)
            else -> result.notImplemented()
        }
    }

    private fun recognise(call: MethodCall, result: MethodChannel.Result) {
        val path = call.argument<String>("path")
        if (path.isNullOrBlank()) {
            result.error("bad_arguments", "recognise requires a 'path'", null)
            return
        }

        val file = File(path)
        if (!file.exists()) {
            result.error("unreadable_image", "No image at $path", null)
            return
        }

        // The image dimensions are needed to normalise bounding boxes, and
        // decoding bounds only is far cheaper than decoding the bitmap.
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(path, bounds)
        val imageWidth = bounds.outWidth.toDouble()
        val imageHeight = bounds.outHeight.toDouble()
        if (imageWidth <= 0 || imageHeight <= 0) {
            result.error("unreadable_image", "Could not read the image size", null)
            return
        }

        val image = try {
            InputImage.fromFilePath(context, Uri.fromFile(file))
        } catch (e: Exception) {
            result.error("unreadable_image", e.message, null)
            return
        }

        recogniser.process(image)
            .addOnSuccessListener { text ->
                val blocks = mutableListOf<Map<String, Any>>()
                for (block in text.textBlocks) {
                    for (line in block.lines) {
                        val box = line.boundingBox ?: continue
                        blocks.add(
                            mapOf(
                                "text" to line.text,
                                // ML Kit reports confidence per element and
                                // not always at all. Where it is absent the
                                // extractor leans on geometry instead, which
                                // is why it must not be faked here.
                                "confidence" to (line.confidence?.toDouble() ?: 1.0),
                                // Normalise to 0..1 with a top-left origin,
                                // matching what Dart expects and what the iOS
                                // side produces after flipping its y axis.
                                "left" to (box.left / imageWidth),
                                "top" to (box.top / imageHeight),
                                "width" to (box.width() / imageWidth),
                                "height" to (box.height() / imageHeight),
                            )
                        )
                    }
                }
                result.success(blocks)
            }
            .addOnFailureListener {
                // Recognition failing is a normal outcome on a worn meter
                // face, not an error the user should see. Dart treats an
                // empty result as "fall back to manual entry".
                result.success(emptyList<Map<String, Any>>())
            }
    }
}
